File: problems/problem_2.py
class Fibonacci:

    def __init__(self, upper_limit):
        self.fibonacci = [1, 2]
        while self.fibonacci[-1] < upper_limit:
            self.fibonacci.append(self.fibonacci[-1] + self.fibonacci[-2])

    def sum(self):
        return sum(self.fibonacci)

    def sum_evens(self):
        return sum([x for x in self.fibonacci if x % 2 == 0])

    def sum_odds(self):
        return sum([x for x in self.fibonacci if x % 2 != 0])

if __name__ == "__main__":
    print(Fibonacci(4000000).sum_evens())

File: problems/euler/src/bin/problem_2.rs
fn main() {
    println!("{}", sum_evens(generate_fibonacci(4000000)));
}

fn sum_evens(numbers: Vec<u64>) -> u64 {
    numbers.iter().filter(|&x| x % 2 == 0).sum()
}

fn generate_fibonacci(upper_limit: u64) -> Vec<u64> {
    let mut fibonacci = vec![1, 2];
    while *fibonacci.last().unwrap() < upper_limit {
        fibonacci.push(fibonacci[fibonacci.len() - 1] + fibonacci[fibonacci.len() - 2]);
    }
    fibonacci
}
